### PR TITLE
[WIP]Read sc name from request

### PR DESCRIPTION
**Pull request by princerachit**
**What this PR does / why we need it**:
- m-apiserver would extract the sc name of pv from request header, if missing it will fetch the pv object then extract the sc name
- Removed default read cas template
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go [hunk 1]
```diff
@@ -139,6 +139,12 @@ func (v *volumeAPIOpsV1alpha1) read(volumeName string) (*v1alpha1.CASVolume, err
 		vol.Namespace = hdrNS
 	}
 
+	// use sc name from header if present
+	scName := v.req.Header.Get(string(v1alpha1.StorageClassKey))
+	if scName != "" {
+		vol.Annotations[string(v1alpha1.StorageClassKey)] = scName
```
Comment (AmitKumarDas) [inline cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go]: This is not available. Remove these lines.
Remember a GET request i.e. read volume has name of volume & http request header

Comment (AmitKumarDas) [reply cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go]: This value needs to be set into the CASVolume object. Do a TrimSpace before setting.

File: pkg/volume/volume.go [hunk 2]
```diff
@@ -215,20 +215,31 @@ func (v *VolumeOperation) Read() (*v1alpha1.CASVolume, error) {
...
-	castName := v.volume.Annotations[string(v1alpha1.CASTemplateKeyForVolumeRead)]
-	if len(castName) == 0 {
-		// use the DEFAULT read cas template otherwise
-		// TODO
-		//  Remove the use of defaults & make volume annotations mandatory
-		// for read operation
-		castName = string(v1alpha1.DefaultCASTemplateForJivaVolumeRead)
+	// check if sc name is already present, if not then extract it
+	scName := v.volume.Annotations[string(v1alpha1.StorageClassKey)]
+	if scName == "" {
```
Comment (AmitKumarDas) [inline pkg/volume/volume.go]: len(scName) == 0

File: pkg/volume/volume.go [hunk 3]
```diff
@@ -215,20 +215,31 @@ func (v *VolumeOperation) Read() (*v1alpha1.CASVolume, error) {
...
+		// fetch the pv specification
+		pv, err := v.k8sClient.GetPV(v.volume.Name, mach_apis_meta_v1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		// extract the sc name
+		scName := pv.Spec.StorageClassName
+		if len(scName) == 0 {
```
Comment (AmitKumarDas) [inline pkg/volume/volume.go]: move this if logic out of the outer if logic.

File: pkg/volume/volume.go [hunk 4]
```diff
@@ -215,20 +215,31 @@ func (v *VolumeOperation) Read() (*v1alpha1.CASVolume, error) {
...
+		}
+	}
+
+	// fetch the sc specification
+	sc, err := v.k8sClient.GetStorageV1SC(scName, mach_apis_meta_v1.GetOptions{})
+	if err != nil {
+		return nil, err
 	}
 
+	// extract read cas template name from sc annotation
```
Comment (AmitKumarDas) [inline pkg/volume/volume.go]: is it read or delete

File: cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go [hunk 5]
```diff
@@ -139,6 +139,12 @@ func (v *volumeAPIOpsV1alpha1) read(volumeName string) (*v1alpha1.CASVolume, err
 		vol.Namespace = hdrNS
 	}
 
+	// use sc name from header if present
+	scName := v.req.Header.Get(string(v1alpha1.StorageClassKey))
+	if scName != "" {
```
Comment (AmitKumarDas) [inline cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go]: len(scName) == 0

File: pkg/client/k8s/k8s.go [hunk 6]
```diff
@@ -101,6 +101,10 @@ type K8sClient struct {
 	// within the current K8s cluster for OpenEBS objects
 	oecs *openebs.Clientset
 
+	// PV refers to a K8s PersistentVolume object
+	// NOTE: This property enables unit testing
```
Comment (avishnu) [inline pkg/client/k8s/k8s.go]: Why this NOTE?

File: pkg/client/k8s/k8s.go [hunk 7]
```diff
@@ -419,6 +423,22 @@ func (k *K8sClient) GetPVC(name string, opts mach_apis_meta_v1.GetOptions) (*api
 	return pops.Get(name, opts)
 }
 
+// coreV1PVOps is a utility function that provides a instance capable of
```
Comment (avishnu) [inline pkg/client/k8s/k8s.go]: Typo - "an instance".

File: pkg/volume/volume.go [hunk 8]
```diff
@@ -215,20 +215,31 @@ func (v *VolumeOperation) Read() (*v1alpha1.CASVolume, error) {
...
-		//  Remove the use of defaults & make volume annotations mandatory
-		// for read operation
-		castName = string(v1alpha1.DefaultCASTemplateForJivaVolumeRead)
+	// check if sc name is already present, if not then extract it
+	scName := v.volume.Annotations[string(v1alpha1.StorageClassKey)]
+	if scName == "" {
+		// fetch the pv specification
+		pv, err := v.k8sClient.GetPV(v.volume.Name, mach_apis_meta_v1.GetOptions{})
+		if err != nil {
+			return nil, err
```
Comment (avishnu) [inline pkg/volume/volume.go]: Is error logged in caller?

File: pkg/volume/volume.go (hunk 4 above)
Comment (princerachit) [reply pkg/volume/volume.go]: It is read

File: pkg/volume/volume.go [hunk 9]
```diff
@@ -215,20 +215,31 @@ func (v *VolumeOperation) Read() (*v1alpha1.CASVolume, error) {
...
+	}
+
+	// fetch the sc specification
+	sc, err := v.k8sClient.GetStorageV1SC(scName, mach_apis_meta_v1.GetOptions{})
+	if err != nil {
+		return nil, err
 	}
 
+	// extract read cas template name from sc annotation
+	castName := sc.Annotations[string(v1alpha1.CASTemplateKeyForVolumeDelete)]
```
Comment (AmitKumarDas) [inline pkg/volume/volume.go]: This makes use of delete. So is it read or delete?

File: pkg/volume/volume.go (hunk 8 above)
Comment (princerachit) [reply pkg/volume/volume.go]: yes

Comment (princerachit) [reply pkg/volume/volume.go]: All errors are logged in the default logger.

**Pull request merged**